### PR TITLE
Fix job status transition

### PR DESCRIPTION
**Pull request by Michaelvll**
Thanks to the observation by @michaelzhiluo (https://github.com/skypilot-org/skypilot/pull/969#issuecomment-1203440396), we found that there is a bug in mismatching of `job_lib.query_job_status` and `job_lib.get_status`.
Previously, we map the ray job status to sky job status with
```
'PENDING': JobStatus.PENDING
'RUNNING': JobStatus.RUNNING
```
This will cause a problem for the job status transition. When a job has not been allocated with required resources, the job will be set to RUNNING in the skylet, due to the following line.
https://github.com/skypilot-org/skypilot/blob/2cb62c542cfc48761d710586bb926f661722ac94/sky/skylet/job_lib.py#L278

**This PR also fixes the race condition caused by periodic job status update by skylet and job status set in generated ray program.**

Tested:
- [x] `tests/run_smoke_tests.sh`

### CODE REVIEW

File: sky/skylet/job_lib.py [hunk 1]
```diff
@@ -41,8 +41,15 @@ def __lt__(self, other):
...
+    # These are intentionally set to one status before, because:
+    # 1. when the ray status indicates the job is PENDING the generated
+    # python program should be not started yet, i.e. the job should be INIT.
+    # 2. when the ray status indicates the job is RUNNING the resources
+    # may not be allocated yet, i.e. the job should be PENDING.
+    # Note: We need to compare the status with the sky cached job status
+    # to determine if the job is actually running.
+    'PENDING': JobStatus.INIT,
+    'RUNNING': JobStatus.PENDING,
     'succeeded': JobStatus.SUCCEEDED,
```
Comment (michaelzhiluo) [inline sky/skylet/job_lib.py:53]: A bit confused, why is not upper case?

File: sky/skylet/job_lib.py [hunk 2]
```diff
@@ -41,8 +41,15 @@ def __lt__(self, other):
 
 
 _RAY_TO_JOB_STATUS_MAP = {
-    'PENDING': JobStatus.PENDING,
-    'RUNNING': JobStatus.RUNNING,
+    # These are intentionally set to one status before, because:
+    # 1. when the ray status indicates the job is PENDING the generated
+    # python program should be not started yet, i.e. the job should be INIT.
```
Comment (michaelzhiluo) [inline sky/skylet/job_lib.py:46]: 1) "Should not be started"
2) What is Sky cached job status?
3) Is it possible to give these Strings a different name? It is easy to confuse `Pending` and Jobstatus's Pending.

File: sky/skylet/job_lib.py (hunk 1 above)
Comment (Michaelvll) [reply sky/skylet/job_lib.py:62]: It is the problem of ray's output. They returns the status in different cases...

File: sky/skylet/job_lib.py (hunk 2 above)
Comment (Michaelvll) [reply sky/skylet/job_lib.py]: Fixed the issues! Thanks! The strings represents the output of `ray job status` we cannot directly change them, unless patching the ray code.

Comment (michaelzhiluo) [reply sky/skylet/job_lib.py]: Ok I think file a Git issue on Ray for the lower case status.

Comment (Michaelvll) [reply sky/skylet/job_lib.py]: That is no longer a problem in the newer version of ray, such as 1.13. ; ) You will see them all in upper case in #969

Comment (michaelzhiluo) [reply sky/skylet/job_lib.py]: Sound good!

File: sky/skylet/log_lib.py [hunk 3]
```diff
@@ -377,7 +377,7 @@ def tail_logs(job_owner: str,
     log_path = os.path.join(log_dir, 'run.log')
     log_path = os.path.expanduser(log_path)
 
-    status = job_lib.query_job_status(job_owner, [job_id])[0]
+    status = job_lib.get_status(job_id)
```
Comment (concretevitamin) [inline sky/skylet/log_lib.py:380]: Why this change? IIUC, `query_job_status()` ensures latest statuses, while `get_status()` is looking up the DB, which may be stale?

File: sky/skylet/log_lib.py [hunk 4]
```diff
@@ -399,7 +399,7 @@ def tail_logs(job_owner: str,
         print(f'INFO: Waiting {_SKY_LOG_WAITING_GAP_SECONDS}s for the logs '
               'to be written...')
         time.sleep(_SKY_LOG_WAITING_GAP_SECONDS)
-        status = job_lib.query_job_status(job_owner, [job_id])[0]
+        status = job_lib.get_status(job_id)
```
Comment (concretevitamin) [inline sky/skylet/log_lib.py:402]: Same question.

**Review by concretevitamin — COMMENTED** [review]
This is a tricky case @Michaelvll @michaelzhiluo.  Some questions.

File: sky/skylet/job_lib.py [hunk 5]
```diff
@@ -24,12 +24,21 @@
 class JobStatus(enum.Enum):
     """Job status"""
     # 3 in-flux states: each can transition to any state below it.
+    # The `job_id` is fetched, but the generated ray program is not
```
[the remark this answers is not in the document]
Comment (Michaelvll) [reply sky/skylet/job_lib.py]: Good point! Fixed. Thanks!

File: sky/skylet/job_lib.py [hunk 6]
```diff
@@ -41,8 +50,15 @@ def __lt__(self, other):
...
 _RAY_TO_JOB_STATUS_MAP = {
-    'PENDING': JobStatus.PENDING,
-    'RUNNING': JobStatus.RUNNING,
+    # These are intentionally set to one status before, because:
+    # 1. when the ray status indicates the job is PENDING the generated
+    # python program should not be started yet, i.e. the job should be INIT.
+    # 2. when the ray status indicates the job is RUNNING the resources
+    # may not be allocated yet, i.e. the job should be PENDING.
+    # Note: We need to compare the status with the job status in our
+    # database for job, to determine if the job is actually running.
```
[the remark this answers is not in the document]
Comment (Michaelvll) [reply sky/skylet/job_lib.py]: Yes, the understanding is correct. I refined the sentence accordingly.

File: sky/skylet/job_lib.py [hunk 7]
```diff
@@ -224,7 +240,10 @@ def _get_jobs_by_ids(job_ids: List[int]) -> List[Dict[str, Any]]:
 
 
 def query_job_status(job_owner: str, job_ids: List[int]) -> List[JobStatus]:
-    """Return the status of the jobs based on the `ray job status` command.
+    """Return the statuses of the jobs.
+
+    This function returns the job status that matches the meaning of each
+    status in the comment of `class JobStatus`.
```
[the remark this answers is not in the document]
Comment (Michaelvll) [reply sky/skylet/job_lib.py]: Good point! I revised the comment. Thank you!

File: sky/skylet/job_lib.py [hunk 8]
```diff
@@ -273,8 +292,12 @@ def query_job_status(job_owner: str, job_ids: List[int]) -> List[JobStatus]:
         else:
             ray_status = res.rpartition(' ')[-1]
             status = _RAY_TO_JOB_STATUS_MAP[ray_status]
-            # To avoid race condition, where the original status has already
+            # Taking max of the status is necessary because:
+            # 1. It avoids race condition, where the original status has already
             # been set to later state by the job. We skip the update.
+            # 2. We map the `ray job status` from RUNNING to PENDING, as the
+            # actual status should be decided by the job status in the database
+            # (set by the generated ray program actively).
```
[the remark this answers is not in the document]
Comment (Michaelvll) [reply sky/skylet/job_lib.py]: Fixed. Thanks!

File: sky/skylet/log_lib.py (hunk 3 above)
Comment (Michaelvll) [reply sky/skylet/log_lib.py:380]: Ahh, that is a good point. I added the job status update back. During adding this, I realized that there could be a race condition between the job status query (or called update now), and the generated ray program. I added a filelock for each job, so that the job status update will be atomic. PTAL.

File: sky/skylet/job_lib.py [hunk 9]
```diff
@@ -239,11 +261,14 @@ def _get_jobs_by_ids(job_ids: List[int]) -> List[Dict[str, Any]]:
     return records
 
 
-def query_job_status(job_owner: str, job_ids: List[int]) -> List[JobStatus]:
-    """Return the statuses of the jobs.
+def update_job_status(job_owner: str,
+                      job_ids: List[int],
+                      need_output: bool = True) -> List[JobStatus]:
```
Comment (concretevitamin) [inline sky/skylet/job_lib.py:266]: nit: `need_output` -> `silent`? The former suggests whether there's any returned value.

File: sky/skylet/log_lib.py [hunk 10]
```diff
@@ -377,7 +379,10 @@ def tail_logs(job_owner: str,
     log_path = os.path.join(log_dir, 'run.log')
     log_path = os.path.expanduser(log_path)
 
-    status = job_lib.get_status(job_id)
+    job_lib.update_job_status(job_owner, [job_id], need_output=False)
+    # No need to lock the status here, as the while loop can handle the
+    # older status.
+    status = job_lib.get_status_no_lock(job_id)
```
Comment (concretevitamin) [inline sky/skylet/log_lib.py:385]: How about making update_job_status() return the values that just got set inside? Same for L408.

File: tests/test_smoke.py [hunk 11]
```diff
@@ -150,6 +150,24 @@ def test_region():
     run_one_test(test)
 
 
+def test_stale_job():
```
Comment (concretevitamin) [inline tests/test_smoke.py:153]: Nice!

File: sky/skylet/job_lib.py [hunk 12]
```diff
@@ -147,7 +152,24 @@ def set_status(job_id: int, status: JobStatus) -> None:
...
+    # TODO(mraheja): remove pylint disabling when filelock version updated.
+    # pylint: disable=abstract-class-instantiated
+    with filelock.FileLock(_JOB_STATUS_LOCK.format(job_id)):
+        _CURSOR.execute(
+            'UPDATE jobs SET status=(?), start_at=(?), end_at=NULL '
+            'WHERE job_id=(?)', (JobStatus.RUNNING.value, time.time(), job_id))
+        _CONN.commit()
+
+
+def get_status_no_lock(job_id: int) -> JobStatus:
```
Comment (concretevitamin) [inline sky/skylet/job_lib.py:172]: Add docstr saying: this may return a stale status if there's a concurrent update. Use get_status() to protect against that?

I still lean towards we should not expose this and used outside; but if we do, the docstring should give some guidance on when to call which func.

File: sky/backends/cloud_vm_ray_backend.py [hunk 13]
```diff
@@ -69,6 +69,7 @@
 
 # The maximum retry count for fetching head IP address.
 _HEAD_IP_MAX_ATTEMPTS = 5
+_WORKER_IP_MAX_ATTEMPTS = 5
```
Comment (concretevitamin) [inline sky/backends/cloud_vm_ray_backend.py:72]: nit: L70 remove "head"

**Review by concretevitamin — COMMENTED** [review]
Thanks @Michaelvll!  A few comments.

File: sky/skylet/job_lib.py [hunk 14]
```diff
@@ -128,14 +136,11 @@ def add_job(job_name: str, username: str, run_timestamp: str,
     return job_id
 
 
-def set_status(job_id: int, status: JobStatus) -> None:
-    assert status != JobStatus.RUNNING, (
```
[the remark this answers is not in the document]
Comment (Michaelvll) [reply sky/skylet/job_lib.py:116]: I don't think we should keep this, as the `update_job_status` can set the status to be RUNNING again. Added a docstr to make sure that we should call `set_started` for the first time setting the job to be RUNNING.

File: sky/skylet/job_lib.py [hunk 15]
```diff
@@ -20,15 +22,17 @@
 
 SKY_LOGS_DIRECTORY = '~/sky_logs'
 
+_JOB_STATUS_LOCK = os.path.expanduser('~/.sky/.sky_job_lock_for_{}.lock')
```
[the remark this answers is not in the document]
Comment (Michaelvll) [reply sky/skylet/job_lib.py:25]: Good point! Moved the lock to `~/.sky/locks`

File: sky/skylet/job_lib.py [hunk 16]
```diff
@@ -263,9 +283,12 @@ def _get_jobs_by_ids(job_ids: List[int]) -> List[Dict[str, Any]]:
 
 def update_job_status(job_owner: str,
                       job_ids: List[int],
-                      need_output: bool = True) -> List[JobStatus]:
-    """Updates the "true" job statuses from our perspective in the database,
-    i.e., with the semantics documented in the comments of `class JobStatus`.
+                      silent: bool = False) -> List[JobStatus]:
+    """Updates and returns the true job statuses that match our semantics
```
Comment (concretevitamin) [inline sky/skylet/job_lib.py:287]: styleguide nit: need to fit on 1 line

maybe something like: Updates and returns true job statuses matching our JobStatus semantics.

File: sky/skylet/job_lib.py [hunk 17]
```diff
@@ -342,8 +365,10 @@ def update_job_status(job_owner: str,
                 # that value. So we take the max here to keep it at RUNNING.
                 status = max(status, original_status)
                 _set_status_no_lock(job_id, status)
-            if need_output:
-                logger.info(f'Updated job {job_id} status to {status}')
+        if not silent:
+            logger.info(f'Updated job {job_id} status to {status}')
```
Comment (concretevitamin) [inline sky/skylet/job_lib.py:369]: If L339's if is not true, we don't really "update" the original status to status (L338), can we skip this logging?

File: sky/skylet/job_lib.py [hunk 18]
```diff
@@ -137,6 +145,11 @@ def add_job(job_name: str, username: str, run_timestamp: str,
 
 
 def _set_status_no_lock(job_id: int, status: JobStatus) -> None:
+    """Setting the status of the job in the database.
+
+    Call the `set_job_stated` at the first time setting the job status
+    to RUNNING so as to update the start_at field.
```
Comment (concretevitamin) [inline sky/skylet/job_lib.py:151]: I think we should add back the assert  as a safer programmatic check. See another comment for a proposal?

File: sky/skylet/job_lib.py [hunk 19]
```diff
@@ -342,8 +365,10 @@ def update_job_status(job_owner: str,
                 # that value. So we take the max here to keep it at RUNNING.
                 status = max(status, original_status)
                 _set_status_no_lock(job_id, status)
```
Comment (concretevitamin) [inline sky/skylet/job_lib.py:367]: To ensure we only ever update to RUNNING once, how about we change these to
```
if status != original_status:  # Prevents redundant update.
    _set_status_no_lock(job_id, max(status, original_status))
```

Comment (Michaelvll) [reply sky/skylet/job_lib.py]: That is a good point! I added the assertion back the `set_status` function. Thanks!

File: sky/skylet/job_lib.py (hunk 17 above)
Comment (Michaelvll) [reply sky/skylet/job_lib.py]: Good catch! Moved the logging back to right place. Thanks!

**Review by concretevitamin — APPROVED** [review]
LGTM
